pkg/osvscanner/customgitignore/dir_test.go

```diff
@@ -0,0 +1,309 @@
+// TODO: check if there's a way to make this: `package customgitignore` instead
+package customgitignore_test
+
+import (
+	"os"
+	"path/filepath"
+	"reflect"
+	"slices"
+	"testing"
+
+	"github.com/google/osv-scanner/pkg/osvscanner/customgitignore"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/format/gitignore"
+)
+
+func TestGitignoreFilesFromIgnoredDir(t *testing.T) {
+	t.Parallel()
+
+	// Create a specific git repo with .gitignore files
+	gitRepo := setupGitRepo(t)
+
+	// Read this dir-tree using customgitignore, starting at
+	// the top of the tree at ./dir_a/dir_b
+	start := filepath.Join(gitRepo, "dir_a", "dir_b")
+
+	// Read this dir-tree using customgitignore, starting at the root
+	patterns, _, err := customgitignore.ParseGitIgnores(start, true)
+	if err != nil {
+		t.Errorf("could not read gitignore patterns for test: %v", err)
+	}
+
+	var hasMatch bool
+
+	// expect ./dir_a/dir_b/.gitignore to be processed
+	//
+	// Normally the `dir_a/dir_b/.gitignore` entry in
+	// REPO_ROOT/.gitignore would stop this .gitignore from being
+	// processed, but it's read because the file has been explicitly
+	// supplied from the command-line.
+	hasMatch = slices.ContainsFunc(patterns, func(p gitignore.Pattern) bool {
```

Comment:
I wonder if these might be better handled by using snapshots - you could probably replace all of the checks with just a single `testutility.NewSnapshot().MatchJSON(t, patterns)` (idk what would give nicer output `MatchText` or `MatchJSON` - I'll let you check and decide)

Reply:
I'm guessing not. `p` is a data structure that's handled in a particular way — I've only bothered to learn how the API works, and that returns a bool.

For reference a pattern looks something like this:


```
&gitignore.pattern{domain:[]string{".", "dir_a", "dir_b"}, pattern:[]string{"DIR_B_GITIGNORE"}, inclusion:false, dirOnly:false, isGlob:false} 
```